test/Microsoft.AspNetCore.Routing.Tests/EndpointFactory.cs

```diff
@@ -21,17 +24,40 @@ internal static class EndpointFactory
             params object[] metadata)
         {
             var d = new List<object>(metadata ?? Array.Empty<object>());
+
+            var routePattern = RoutePatternFactory.Parse(template, defaults, policies);
+
             if (requiredValues != null)
             {
-                d.Add(new RouteValuesAddressMetadata(new RouteValueDictionary(requiredValues)));
+                var policyFactory = CreateParameterPolicyFactory();
+                var defaultRoutePatternTransformer = new DefaultRoutePatternTransformer(policyFactory);
+
+                routePattern = defaultRoutePatternTransformer.SubstituteRequiredValues(routePattern, requiredValues);
```

Comment:
Do we need to do this here? The pattern can be constructed with required values, but it will validate and throw if it's invalid. It's is **really** important that route patterns can be constructed without services.

Reply:
Moved this logic to the few integration-style tests that use it to replace defaults.